desktop/src/main/local-protocol.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * Provides secure access to local media files by rewriting `file://`
+ * URLs to use this custom protocols which requires a HMAC
+ * (hash-based message authentication code) for each requested path.
```

Comment:
The HMAC is just a hash of the file path using a secret salt. So if an attacker tried to use a different path, say one that included path traversal, the HMAC would not match and the request would be refused. I'm pretty confident that this approach will only grant access to files for which a HMAC has been generated although some tests for this should be added.

Currently the file paths are absolute. It would be good to make them relative to some directory and have an additional check against path traversal but then we need to define which directory that we are checking for traversal out of (perhaps the project directory, or the parent directory of the document?) The problem then would become how to make the `local://` protocol aware what the root directory is for each request (it will change for each document). My fear here is that we end up making things less secure by introducing more complexity but will have a think about it.

Reply:
I've implemented restricting file links (i.e. references) to be within the project in 4c81020356ccf9a725537c57cd4aa8507574e0b2. This is at the lowest level, when a document is compiled in Rust. I think it is safer to assert that any file links be intra-project, both from the security and reproducibility standpoint, than to only apply it when "serving" files within Electron.